### ISSUE TITLE
Type variable scoping with multiple calls of same functions are not independent.

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4852]
### Discussed in https://github.com/microsoft/pyright/discussions/4847

<div type='discussions-op-text'>

<sup>Originally posted by **sylee957** March 27, 2023</sup>
```python3
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")
A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
X = TypeVar("X")
Y = TypeVar("Y")
Z = TypeVar("Z")


def identity(x: T) -> T:
    return x


def triple(
    f: Callable[[A], X], g: Callable[[B], Y], h: Callable[[C], Z]
) -> Callable[[A, B, C], tuple[X, Y, Z]]:
    def wrapped(a: A, b: B, c: C) -> tuple[X, Y, Z]:
        return f(a), g(b), h(c)

    return wrapped
```

And if I want to narrow down `triple` to `(A, B, C) -> (X, B, C)` function by passing the identity functions, it doesn't work as intended.
because the type variables associated with multiple calls of same functions are somewhat recognized as dependent.

```python
def test_1(f: Callable[[A], X]) -> Callable[[A, B, C], tuple[X, B, C]]:
    return triple(f, identity, identity)  # type check fails
```

However, if I define the duplicate identity functions, it works.

```python3
def identity_1(x: T) -> T:
    return identity(x)


def identity_2(x: T) -> T:
    return identity(x)


def test_2(f: Callable[[A], X]) -> Callable[[A, B, C], tuple[X, B, C]]:
    return triple(f, identity_1, identity_2)
```

I'm not sure if this is a bug or intended behavior.
I'm using pyright version `1.1.300`</div>

**Comment by erictraut** [comment]
This turns out to be a difficult problem to solve because of the way TypeVars are tracked internally to pyright. Interestingly, mypy has the same problem (generates the same error) even though it uses a very different way to track TypeVars internally.

```python
from typing import TypeVar, Callable

T = TypeVar("T")
A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
X = TypeVar("X")
Y = TypeVar("Y")
Z = TypeVar("Z")


def identity(x: T) -> T:
    return x


def triple(
    f: Callable[[A], X], g: Callable[[B], Y], h: Callable[[C], Z]
) -> Callable[[A, B, C], tuple[X, Y, Z]]:
    def wrapped(a: A, b: B, c: C) -> tuple[X, Y, Z]:
        return f(a), g(b), h(c)

    return wrapped


def test_1(f: Callable[[A], X]) -> Callable[[A, B, C], tuple[X, B, C]]:
    xxx = triple(f, identity, identity)

    # Pyright: Type of "xxx" is "(A@test_1, T@identity, T@identity) -> tuple[X@test_1, T@identity, T@identity]"
    # Mypy: Revealed type is "def (A`-1, T`-1, T`-1) -> Tuple[X`-2, T`-1, T`-1]"
    reveal_type(xxx)

    # Pyright: Expression of type "(A@test_1, T@identity, T@identity) -> tuple[X@test_1, T@identity, T@identity]" cannot be assigned to return type "(A@test_1, B@test_1, C@test_1) -> tuple[X@test_1, B@test_1, C@test_1]"
    # Mypy: Incompatible return value type (got "Callable[[A, T, T], Tuple[X, T, T]]", expected "Callable[[A, B, C], Tuple[X, B, C]]")
    return xxx
```

It appears that TypeScript solves this problem by synthesizing a new TypeVar under a different name (i.e. by appending a "1") to differentiate between the two instances of `identity`.

```typescript
export function triple<A, B, C, X, Y, Z>(
    f: (v1: A) => X,
    g: (v1: B) => Y,
    h: (v1: C) => Z
): (v1: A, v2: B, v3: C) => [X, Y, Z] {
    function inner(v1: A, v2: B, v3: C): [X, Y, Z] {
        return [f(v1), g(v2), h(v3)];
    }

    return inner;
}

export function identity<T>(x: T): T {
    return x;
}

export function test_1<A, B, C, X>(f: (v1: A) => X): (v1: A, v2: B, v3: C) => [X, B, C] {
    const xxx = triple(f, identity, identity);
    return xxx;
}
```

![image](https://user-images.githubusercontent.com/7040122/229219642-808ac3d2-816d-423f-9af2-b4e660c1ba5a.png)

A similar approach could work here. I'll need to give it more thought.

**Comment by erictraut** [comment]
I found a way to make this work similar to TypeScript. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.302, which I just published. It will also be included in a future release of pylance.
